presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -594,6 +600,59 @@ private static Split createRemoteSplitFor(TaskId taskId, URI remoteSourceTaskLoc
         return new Split(REMOTE_CONNECTOR_ID, new RemoteTransactionHandle(), new RemoteSplit(new Location(splitLocation), remoteSourceTaskId));
     }
 
+    private String getCteIdFromSource(PlanNode source)
+    {
+        // Traverse the plan node tree to find a TableWriterNode with TemporaryTableInfo
+        return PlanNodeSearcher.searchFrom(source)
+                .where(planNode -> planNode instanceof TableFinishNode)
+                .findFirst()
+                .map(planNode -> ((TableFinishNode) planNode).getTemporaryTableInfo().orElseThrow(
+                        () -> new IllegalStateException("TableFinishNode has no TemporaryTableInfo")))
+                .map(TemporaryTableInfo::getCteId)
+                .orElseThrow(() -> new IllegalStateException("TemporaryTableInfo has no CTE ID"));
+    }
+
+    public boolean isCTETableFinishStage()
+    {
+        return PlanNodeSearcher.searchFrom(planFragment.getRoot())
+                .where(planNode -> (planNode instanceof TableFinishNode))
+                .findAll().stream()
```

Comment:
do you think we should verify there is only one `TableFinishNode` using something like `getOnlyElement`?

Reply:
you can use findSingle() for this. Also, instead of streaming again and then anyMatch, you can move all of that into the where condition above.